MassSpectrometry/MzSpectra/MzSpectrum.cs

```diff
@@ -228,109 +229,256 @@ public override string ToString()
-                    double totalIntensity = candidateForMostIntensePeakIntensity;
-                    for (int indexToLookAt = 1; indexToLookAt < allIntensities[massIndex].Length; indexToLookAt++)
+                    //Find what charge states this peak might be based on the spacing of nearby peaks (assumes isotopic resolution)
+                    HashSet<int> allPossibleChargeStates = new HashSet<int>();
+                    for (int i = candidateForMostIntensePeakIndex + 1; i < XArray.Length; i++) //look at peaks of higher m/z
                     {
-                        //Console.WriteLine("   indexToLookAt: " + indexToLookAt);
-                        double theorMassThatTryingToFind = allMasses[massIndex][indexToLookAt] + differenceBetweenTheorAndActual;
-                        //Console.WriteLine("   theorMassThatTryingToFind: " + theorMassThatTryingToFind);
-                        //Console.WriteLine("   theorMassThatTryingToFind.ToMz(chargeState): " + theorMassThatTryingToFind.ToMz(chargeState));
-                        var closestPeakToTheorMass = GetClosestPeakIndex(theorMassThatTryingToFind.ToMz(chargeState));
-                        var closestPeakmz = XArray[closestPeakToTheorMass.Value];
-                        //Console.WriteLine("   closestPeakmz: " + closestPeakmz);
-                        var closestPeakIntensity = YArray[closestPeakToTheorMass.Value];
-                        if (Math.Abs(closestPeakmz.ToMass(chargeState) - theorMassThatTryingToFind) / theorMassThatTryingToFind * 1e6 <= deconvolutionTolerancePpm
-                            && Peak2satisfiesRatio(allIntensities[massIndex][0], allIntensities[massIndex][indexToLookAt], candidateForMostIntensePeakIntensity, closestPeakIntensity, intensityRatioLimit)
-                            && !listOfPeaks.Contains((closestPeakmz, closestPeakIntensity)))
+                        double deltaMass = XArray[i] - candidateForMostIntensePeakMz;
+                        if (deltaMass < 1.1) //if we're past a Th spacing, we're no longer looking at the closest isotope
                         {
-                            //Found a match to an isotope peak for this charge state!
-                            //Console.WriteLine(" *   Found a match to an isotope peak for this charge state!");
-                            //Console.WriteLine(" *   chargeState: " + chargeState);
-                            //Console.WriteLine(" *   closestPeakmz: " + closestPeakmz);
-                            listOfPeaks.Add((closestPeakmz, closestPeakIntensity));
-                            totalIntensity += closestPeakIntensity;
-                            listOfRatios.Add(allIntensities[massIndex][indexToLookAt] / closestPeakIntensity);
+                            //get the lower bound charge state
+                            int charge = (int)Math.Floor(1 / deltaMass); //e.g. deltaMass = 0.4 Th, charge is now 2 (but might be 3)
```

Comment:
Our deconvolution tolerance in default is 4 ppm. So i don't think this will happen.

Reply:
Sure. But I think it's important to include because:
1. We might change the tolerance someday
2. It does affect high MW species and things at high charge states. 50kDa protein at 4 ppm is +-0.2 Da. Difference between charge 9 and charge 10 is 0.11111-0.1 = 0.0111 Th